src/rpc/protocol.cpp

```diff
@@ -166,16 +180,32 @@ int ReadHTTPStatus(std::basic_istream<char>& stream, int &proto)
     boost::split(vWords, str, boost::is_any_of(" "));
     if (vWords.size() < 2)
         return HTTP_INTERNAL_SERVER_ERROR;
-    proto = 0;
-    const char *ver = strstr(str.c_str(), "HTTP/1.");
-    if (ver != nullptr && !ParseInt(std::string(ver+7), &proto)) {
-        throw std::invalid_argument("Unable to parse proto.");
+
+    size_t length = str.length();
+
+    // If CR is found at the end of the string, strip it, because ParseInt will fail if it is present.
+    if (str.substr(length - 1) == "\r") {
+        str = str.substr(0, length - 1);
+        --length;
+    }
+
+    size_t start_pos = str.find("HTTP/1.");
+
+    if (start_pos != std::string::npos && str.length() - start_pos > 7) {
+        str = str.substr(start_pos + 7);
+
+        if (!ParseInt(str, &proto)) {
+            error("%s: Unable to parse protocol in HTTP string: %s", __func__, str);
+            return HTTP_INTERNAL_SERVER_ERROR;
```

Comment:
```suggestion
            return -1;
```
nit: Server sending an invalid response should not be equivalent to a proper internal server error. Will need a tweak in `rpc/client.cpp`.

Reply:
Throwing an exception could also work without a tweak.

Reply:
If we throw an exception, it will terminate the entire application unless we handle it appropriately. I don't think that is the appropriate response. The original atoi is far more permissive, as you know and would just return zero, which is equivalent to empty brackets inside the if statement here. Maybe that is what we should do...

Reply:
The only caller (`CallRPC`) already throws `runtime_error` on error. It also continues on HTTP_INTERNAL_SERVER_ERROR, for some reason.

Reply:
Fine then we can throw an exception.

Reply:
Actually, considering the original code ignored parsing errors (i.e. just returned zero for protocol), I think we should log an error to the debug log, but not throw an exception at all in either this case or the below. That is at least better than the original code.

Reply:
This was the original code...

    // parse proto, if present
    std::string strProto;
    if (vWords.size() > 2)
        strProto = vWords[2];

    proto = 0;
    const char *ver = strstr(strProto.c_str(), "HTTP/1.");
    if (ver != nullptr)
        proto = atoi(ver+7);

    return true;